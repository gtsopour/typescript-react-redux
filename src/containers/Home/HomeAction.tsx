import {Action, Dispatch} from 'redux';
import {Project} from '../../models/Project';
import {projectsService} from '../../services/ProjectsService';

export enum HomeActionType {
  FETCH_PROJECTS = 'FETCH_PROJECTS',
  FETCH_PROJECTS_ERROR = 'FETCH_PROJECTS_ERROR',
  RESET_PROJECTS = 'RESET_PROJECTS'
}

export interface HomeAction extends Action {
  type: HomeActionType;
  payload: Project[];
}

export const fetchProjects: any = () => {
  return (dispatch: Dispatch) => {
    projectsService.fetchProjects()
    .then((response: any) => {
      dispatch({ type: HomeActionType.FETCH_PROJECTS, payload: response });
    })
    .catch(error => {
      dispatch({ type: HomeActionType.FETCH_PROJECTS_ERROR, payload: error });
    });
  };
};

export const resetProjects: any = () => {
  return (dispatch: Dispatch) => {
    dispatch({ type: HomeActionType.RESET_PROJECTS, payload: [] });
  };
};
