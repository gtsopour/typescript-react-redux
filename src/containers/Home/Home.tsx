import * as React from 'react';
import './styles.scss';
import {connect} from 'react-redux';
import {ActionCreatorsMapObject, bindActionCreators, Dispatch} from 'redux';
import {fetchProjects, resetProjects} from './HomeAction';
import {Project} from '../../models/Project';

interface Props {
  fetchProjects: () => void;
  resetProjects: () => void;
  projects: Project[];
}

interface State {}

class Home extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
  }

  componentDidMount() {
    const { fetchProjects, resetProjects } = this.props;
    resetProjects();
    fetchProjects();
  }

  render() {
    const { projects } = this.props;
    const projectsList = projects.map((item, key) =>
      <div key={key}>({item.id}) {item.name}</div>
    );
    return (
      <div>{projectsList}</div>
    )
  }
}

const mapDispatchToProps: ((dispatch: Dispatch) => ActionCreatorsMapObject) = (dispatch) => {
  return bindActionCreators({ fetchProjects, resetProjects }, dispatch);
};

const mapStateToProps: any = ({ homeReducer }: any) => {
  return {
    projects: homeReducer.projects,
    projectsError: homeReducer.projectsError
  };
};

export default connect<Props>(mapStateToProps, mapDispatchToProps)(Home);
