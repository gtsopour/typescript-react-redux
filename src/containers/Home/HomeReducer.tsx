import {Project} from '../../models/Project';
import {HomeAction, HomeActionType} from './HomeAction';

export interface InitialState {
  projects: Project[] | null;
}

const initialState: InitialState = {
  projects: []
};

export default (state = initialState, action: HomeAction) => {
  switch (action.type) {
    case HomeActionType.FETCH_PROJECTS:
      return {
        ...state,
        projects: action.payload,
      };
    case HomeActionType.FETCH_PROJECTS_ERROR:
      return {
        ...state,
        projectsError: action.payload
      };
    case HomeActionType.RESET_PROJECTS:
      return {
        ...state,
        projects: action.payload
      };
    default:
      return state;
  }
};
