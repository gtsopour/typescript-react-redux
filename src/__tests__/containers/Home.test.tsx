import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import Home from '../../containers/Home/Home';
import reducer from '../../reducers/Reducer';

describe('Home component tests', () => {
  it('Check Home Component Render', () => {
    const store = createStore(reducer, applyMiddleware(thunk));
    const component = renderer.create(<Provider store={store}><Home /></Provider>).toJSON();
    expect(component).toMatchSnapshot();
  });
});
