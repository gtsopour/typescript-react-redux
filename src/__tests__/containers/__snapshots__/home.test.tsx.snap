// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component tests Check Home Component Render 1`] = `<div />`;
